### ISSUE TITLE
Memory leaks when training on colab

### ISSUE THREAD

**Issue by sunnyqin98** [YeongHyeon/DGM-TF#2]
Hi! I am trying to run the code on colab but unfortunately, there is a memory leak during training. Since the default version of TensorFlow of colab is v2, I have to set it back into v1 so that it runs successfully.  Could you please provide the versions of the packages of your environment for this repo? Thanks in advance!

**Comment by YeongHyeon** [comment]
@sunnyqin98 
Thank you for your interest.
I will confirm this issue in Colab environment.

However, it may take some time for solving due to other tasks.
I will notify you when it is processed.
Thank you.

Reference: https://github.com/YeongHyeon/DGM-TF/issues/1

**Comment by YeongHyeon** [comment]
I am using the following command to check memory usage.

``` python
print("RAM:", memory_usage_percent)
```

In script <a href="https://github.com/YeongHyeon/DGM-TF/blob/master/source/tf_process.py">tf_process.py</a>, the above command is added as following.
``` python
import psutil

def training(neuralnet, dataset, epochs, batch_size, normalize=True):

    print("\nTraining to %d epochs (%d of minibatch size)" %(epochs, batch_size))

    make_dir(path="training")
    result_list = ["restoration"]
    for result_name in result_list: make_dir(path=os.path.join("training", result_name))

    iteration = 0
    test_sq = 20
    test_size = test_sq**2
    for epoch in range(epochs):

        x_tr, y_tr, _ = dataset.next_train(batch_size=test_size, fix=True) # Initial batch
        step_dict = neuralnet.step(x=x_tr, y=y_tr, training=False)
        y_hat = step_dict['y_hat']
        save_img(contents=[x_tr, y_hat, (x_tr-y_hat)**2], \
            names=["Input\n(x)", "Restoration\n(x to x-hat)", "Difference"], \
            savename=os.path.join("training", "restoration", "%08d.png" %(epoch)))

        while(True):
            x_tr, y_tr, terminator = dataset.next_train(batch_size)
            step_dict = neuralnet.step(x=x_tr, y=y_tr, iteration=iteration, training=True)

            iteration += 1
            if(terminator): break

        memory_usage_dict = dict(psutil.virtual_memory()._asdict())
        memory_usage_percent = memory_usage_dict['percent']
        print("RAM:", memory_usage_percent)
        print("Epoch [%d / %d] (%d iteration) | G:%.3f, D:%.3f" \
            %(epoch, epochs, iteration, step_dict['loss_g'], step_dict['loss_d']))
        neuralnet.save_parameter(model='model_checker', epoch=epoch)
```

First three epochs the memory usage remains below 10%.

``` sh
Training to 100 epochs (32 of minibatch size)
RAM: 8.6
Epoch [0 / 100] (63 iteration) | G:16515.086, D:0.957
RAM: 9.2
Epoch [1 / 100] (126 iteration) | G:11777.293, D:0.495
RAM: 9.1
Epoch [2 / 100] (189 iteration) | G:9254.162, D:0.308
```

When the training is finished, I will check further.

**Comment by YeongHyeon** [comment]
In MacBook pro 15' (with other prosess)
![list_ram](https://user-images.githubusercontent.com/22256220/125789706-ca4e2cd3-e9f9-4cb1-8403-e138fa6c897a.png)

**Comment by sunnyqin98** [comment]
Thank you very much for the instant reply! Currently, I manage to start the training on Colab with TensorFlow 2.5.0  by altering the following line in the encoder and decoder functions. 
`with tf.compat.v1.variable_scope(name, reuse=reuse):`
I'll continue observing the memory loss and hope the memory leak will not happen anymore. The memory leak is more likely caused by some inner issues of colab backend and probably the difference of the tf versions.

**Comment by YeongHyeon** [comment]
@sunnyqin98 
Thank you for informing me.
Refer that, this source code was created in the environment below.
(as shown in <a href="https://github.com/YeongHyeon/DGM-TF">main page</a>.)
* Python 3.7.4
* Tensorflow 1.14.0
* Numpy 1.17.1
* Matplotlib 3.1.1
* Scikit Learn (sklearn) 0.21.3

**Comment by YeongHyeon** [comment]
In Colab with TensorFlow 1.14

![image](https://user-images.githubusercontent.com/22256220/125801597-6d9ebfbe-be9d-404d-bff0-e6fa6bf0d445.png)


<details>
  <summary>Training log (click here) </summary>  

``` sh
Training to 100 epochs (32 of minibatch size)
RAM:8.600%
Epoch [0 / 100] (63 iteration) | G:21808.033, D:0.644
RAM:8.600%
Epoch [1 / 100] (126 iteration) | G:14930.521, D:0.626
RAM:8.600%
Epoch [2 / 100] (189 iteration) | G:10220.294, D:0.603
RAM:9.000%
Epoch [3 / 100] (252 iteration) | G:7788.000, D:0.438
RAM:9.200%
Epoch [4 / 100] (315 iteration) | G:6539.430, D:0.365
RAM:9.200%
Epoch [5 / 100] (378 iteration) | G:5004.999, D:0.265
RAM:9.200%
Epoch [6 / 100] (441 iteration) | G:4771.496, D:0.221
RAM:9.400%
Epoch [7 / 100] (504 iteration) | G:4029.305, D:0.193
RAM:9.500%
Epoch [8 / 100] (567 iteration) | G:3177.107, D:0.193
RAM:9.800%
Epoch [9 / 100] (630 iteration) | G:3382.224, D:0.226
RAM:9.800%
Epoch [10 / 100] (693 iteration) | G:3251.442, D:0.182
RAM:9.800%
Epoch [11 / 100] (756 iteration) | G:2513.884, D:0.185
RAM:9.000%
Epoch [12 / 100] (819 iteration) | G:2787.007, D:0.172
RAM:9.000%
Epoch [13 / 100] (882 iteration) | G:2022.413, D:0.212
RAM:9.200%
Epoch [14 / 100] (945 iteration) | G:2056.981, D:0.231
RAM:9.200%
Epoch [15 / 100] (1008 iteration) | G:2186.975, D:0.206
RAM:9.300%
Epoch [16 / 100] (1071 iteration) | G:2267.356, D:0.168
RAM:9.400%
Epoch [17 / 100] (1134 iteration) | G:1860.046, D:0.197
RAM:9.400%
Epoch [18 / 100] (1197 iteration) | G:1432.073, D:0.183
RAM:9.400%
Epoch [19 / 100] (1260 iteration) | G:1554.894, D:0.215
RAM:9.600%
Epoch [20 / 100] (1323 iteration) | G:1578.873, D:0.218
RAM:9.700%
Epoch [21 / 100] (1386 iteration) | G:1410.038, D:0.228
RAM:9.900%
Epoch [22 / 100] (1449 iteration) | G:1385.650, D:0.254
RAM:9.900%
Epoch [23 / 100] (1512 iteration) | G:1507.188, D:0.246
RAM:9.900%
Epoch [24 / 100] (1575 iteration) | G:1405.552, D:0.229
RAM:10.000%
Epoch [25 / 100] (1638 iteration) | G:1561.428, D:0.219
RAM:10.000%
Epoch [26 / 100] (1701 iteration) | G:1220.580, D:0.264
RAM:10.500%
Epoch [27 / 100] (1764 iteration) | G:1280.036, D:0.282
RAM:10.600%
Epoch [28 / 100] (1827 iteration) | G:1182.446, D:0.270
RAM:10.600%
Epoch [29 / 100] (1890 iteration) | G:1298.184, D:0.221
RAM:10.600%
Epoch [30 / 100] (1953 iteration) | G:756.467, D:0.334
RAM:10.600%
Epoch [31 / 100] (2016 iteration) | G:1037.806, D:0.284
RAM:11.100%
Epoch [32 / 100] (2079 iteration) | G:1052.483, D:0.240
RAM:11.100%
Epoch [33 / 100] (2142 iteration) | G:1032.020, D:0.225
RAM:9.100%
Epoch [34 / 100] (2205 iteration) | G:935.464, D:0.281
RAM:9.100%
Epoch [35 / 100] (2268 iteration) | G:1111.781, D:0.257
RAM:9.100%
Epoch [36 / 100] (2331 iteration) | G:933.646, D:0.258
RAM:9.300%
Epoch [37 / 100] (2394 iteration) | G:1038.284, D:0.246
RAM:9.300%
Epoch [38 / 100] (2457 iteration) | G:815.816, D:0.271
RAM:9.400%
Epoch [39 / 100] (2520 iteration) | G:947.824, D:0.283
RAM:9.400%
Epoch [40 / 100] (2583 iteration) | G:1038.887, D:0.260
RAM:9.400%
Epoch [41 / 100] (2646 iteration) | G:902.050, D:0.261
RAM:9.700%
Epoch [42 / 100] (2709 iteration) | G:1129.969, D:0.268
RAM:9.700%
Epoch [43 / 100] (2772 iteration) | G:1078.139, D:0.263
RAM:9.800%
Epoch [44 / 100] (2835 iteration) | G:1111.610, D:0.186
RAM:9.800%
Epoch [45 / 100] (2898 iteration) | G:952.733, D:0.259
RAM:9.900%
Epoch [46 / 100] (2961 iteration) | G:836.265, D:0.302
RAM:10.000%
Epoch [47 / 100] (3024 iteration) | G:976.528, D:0.276
RAM:10.100%
Epoch [48 / 100] (3087 iteration) | G:930.874, D:0.289
RAM:10.500%
Epoch [49 / 100] (3150 iteration) | G:778.756, D:0.301
RAM:10.500%
Epoch [50 / 100] (3213 iteration) | G:885.353, D:0.284
RAM:10.500%
Epoch [51 / 100] (3276 iteration) | G:877.115, D:0.266
RAM:10.500%
Epoch [52 / 100] (3339 iteration) | G:980.689, D:0.257
RAM:10.600%
Epoch [53 / 100] (3402 iteration) | G:943.214, D:0.247
RAM:11.000%
Epoch [54 / 100] (3465 iteration) | G:966.477, D:0.274
RAM:11.000%
Epoch [55 / 100] (3528 iteration) | G:883.570, D:0.251
RAM:9.100%
Epoch [56 / 100] (3591 iteration) | G:576.102, D:0.326
RAM:9.100%
Epoch [57 / 100] (3654 iteration) | G:769.445, D:0.296
RAM:9.100%
Epoch [58 / 100] (3717 iteration) | G:897.141, D:0.263
RAM:9.200%
Epoch [59 / 100] (3780 iteration) | G:863.670, D:0.276
RAM:9.400%
Epoch [60 / 100] (3843 iteration) | G:919.883, D:0.248
RAM:9.400%
Epoch [61 / 100] (3906 iteration) | G:985.954, D:0.227
RAM:9.400%
Epoch [62 / 100] (3969 iteration) | G:917.939, D:0.262
RAM:9.500%
Epoch [63 / 100] (4032 iteration) | G:732.476, D:0.288
RAM:9.700%
Epoch [64 / 100] (4095 iteration) | G:847.399, D:0.279
RAM:9.800%
Epoch [65 / 100] (4158 iteration) | G:1161.942, D:0.243
RAM:10.000%
Epoch [66 / 100] (4221 iteration) | G:774.432, D:0.305
RAM:10.000%
Epoch [67 / 100] (4284 iteration) | G:879.421, D:0.266
RAM:10.100%
Epoch [68 / 100] (4347 iteration) | G:1042.294, D:0.233
RAM:10.100%
Epoch [69 / 100] (4410 iteration) | G:824.700, D:0.279
RAM:10.100%
Epoch [70 / 100] (4473 iteration) | G:880.509, D:0.239
RAM:10.500%
Epoch [71 / 100] (4536 iteration) | G:917.437, D:0.274
RAM:10.600%
Epoch [72 / 100] (4599 iteration) | G:876.266, D:0.264
RAM:10.700%
Epoch [73 / 100] (4662 iteration) | G:721.387, D:0.287
RAM:10.700%
Epoch [74 / 100] (4725 iteration) | G:853.777, D:0.248
RAM:10.700%
Epoch [75 / 100] (4788 iteration) | G:768.721, D:0.291
RAM:10.900%
Epoch [76 / 100] (4851 iteration) | G:838.282, D:0.257
RAM:11.100%
Epoch [77 / 100] (4914 iteration) | G:807.442, D:0.271
RAM:9.200%
Epoch [78 / 100] (4977 iteration) | G:723.278, D:0.281
RAM:9.200%
Epoch [79 / 100] (5040 iteration) | G:777.948, D:0.260
RAM:9.200%
Epoch [80 / 100] (5103 iteration) | G:825.232, D:0.292
RAM:9.300%
Epoch [81 / 100] (5166 iteration) | G:1004.016, D:0.240
RAM:9.400%
Epoch [82 / 100] (5229 iteration) | G:668.530, D:0.291
RAM:9.500%
Epoch [83 / 100] (5292 iteration) | G:937.316, D:0.264
RAM:9.600%
Epoch [84 / 100] (5355 iteration) | G:866.665, D:0.233
RAM:9.600%
Epoch [85 / 100] (5418 iteration) | G:733.542, D:0.295
RAM:9.700%
Epoch [86 / 100] (5481 iteration) | G:955.487, D:0.225
RAM:9.800%
Epoch [87 / 100] (5544 iteration) | G:784.796, D:0.246
RAM:9.900%
Epoch [88 / 100] (5607 iteration) | G:740.729, D:0.281
RAM:10.100%
Epoch [89 / 100] (5670 iteration) | G:894.222, D:0.256
RAM:10.100%
Epoch [90 / 100] (5733 iteration) | G:742.077, D:0.292
RAM:10.200%
Epoch [91 / 100] (5796 iteration) | G:878.311, D:0.268
RAM:10.200%
Epoch [92 / 100] (5859 iteration) | G:677.531, D:0.299
RAM:10.300%
Epoch [93 / 100] (5922 iteration) | G:873.586, D:0.242
RAM:10.700%
Epoch [94 / 100] (5985 iteration) | G:895.804, D:0.267
RAM:10.700%
Epoch [95 / 100] (6048 iteration) | G:862.353, D:0.259
RAM:10.700%
Epoch [96 / 100] (6111 iteration) | G:630.236, D:0.344
RAM:10.800%
Epoch [97 / 100] (6174 iteration) | G:850.473, D:0.250
RAM:10.700%
Epoch [98 / 100] (6237 iteration) | G:858.242, D:0.268
RAM:11.200%
Epoch [99 / 100] (6300 iteration) | G:870.985, D:0.232
```
</details>

**Comment by YeongHyeon** [comment]
It is true that RAM usage continues to change, but it has been confirmed that there is no leak such as a linear increase.
If the problem when applying to other (your own) datasets, I will check it out further.

Thank you once more and always good luck to you.
:)
